### PR TITLE
Add core programmatic API align with CLI

### LINKED ISSUE #871: Move the functionality of cli to sdk
We need to move the functionality of the current CLI to SDK so that we can align the CLI and programmatic APIs. One example of the benefit is we can replace the `subprocess` call of the CLI commands with SDK, e.g. `subprocess.run('sky status')` in the cloud_vm_ray_backend.

### PR DESCRIPTION
**Pull request by Michaelvll**
Closes #871 and #980. Also, makes #972 easier.

Our current implementation has many functionalities in the `cli.py`, which makes python API users have to use subprocess to do the cluster, jobs, and storage management, and cannot handle easily handle the return values (needs to be done with string process).

This PR currently just has the skeleton of SDK and APIs. The implementation will be filled later. Please take a look at the API hierarchy and let me know if you find it not appropriate. : )

~~I am not sure whether the `stop`, `start`, `down`, `autostop` should be placed under `sky` or `sky.sdk`. If placed under `sky.sdk`, it is a bit weird to only have `launch` and `exec` under `sky`.~~

I updated the hierarchy of the APIs based on the [suggestion](https://github.com/sky-proj/sky/pull/978#issuecomment-1186033792) of @romilbhardwaj.

### Other changes
1. Added a smoke test for the `sky logs`.

### Hierarchy of  the APIs:
**sky**
  - sky.launch
  - sky.exec
  - sky.stop
  - sky.start
  - sky.down
  - sky.autostop
  - sky.status
  - sky.queue
  - sky.cancel
  - sky.spot_status
  - sky.spot_cancel
  - sky.get_storage
  - sky.delete_storage

### TODOs
- [ ] Add docs for python API (#1046 )

#### CLIs
- [x] `sky launch -c min ./example/minimal.yaml`
- [x] `sky stop min`
- [x] `sky start min` 
- [x] `sky autostop min`
- [x] `sky autostop --cancel min`
- [x] `sky queue`
- [x] `sky spot launch -n test-queue 'sleep 10000; echo hi'`
- [x] `sky spot status`
- [x] `sky spot cancel -n test-queue`
- [x] `sky logs test-multi-echo-zhwu-164b 1 2 3 4 -s`
- [x] `sky logs test-multi-echo-zhwu-164b 3 -s`
- [x] `sky logs test-multi-echo-zhwu-164b 3`
- [x] `sky logs test-multi-echo-zhwu-164b 1 2 3 4` fails
- [x] `tests/run_smoke_test.sh` (Failed TPU VM as #962)

### CODE REVIEW

**Comment by Michaelvll** [comment]
Merged the master branch.
Tested:
- [x] `./tests/run_smoke_tests.sh`

TODO:
- [ ] Move the Onprem and Benchmark CLI into sdk.

**Comment by concretevitamin** [comment]
> For the onprem and benchmark, I have not moved them into the core as they may not be immediately needed yet, and may experience some API change in the short future. We can move them in a future PR. Wdyt @romilbhardwaj @concretevitamin ?

This sounds good to me.

File: sky/core.py [hunk 1]
```diff
@@ -0,0 +1,590 @@
...
+    if handle.head_ip is None:
+        raise exceptions.ClusterNotUpError(
+            f'Cluster {cluster_name!r} has been stopped or not properly set up.'
+            ' Please re-launch it with `sky start`.')
+    return handle
+
+
+def queue(cluster_name: str,
+          skip_finished: bool = False,
+          all_users: bool = False):
```
Comment (romilbhardwaj) [inline sky/core.py:228]: Add return type `-> List[dict]`?

File: sky/data/storage_utils.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
+"""Utility functions for the storage module."""
```
Comment (romilbhardwaj) [inline sky/data/storage_utils.py:1]: Thanks for creating this!

**Review by romilbhardwaj — COMMENTED** [review]
This is great! We will have a much cleaner python API after this PR. Left some comments

File: sky/backends/cloud_vm_ray_backend.py [hunk 3]
```diff
@@ -2046,69 +2047,74 @@ def cancel_jobs(self, handle: ResourceHandle, jobs: Optional[List[int]]):
...
+        if not log_dirs:
             logger.info(f'{colorama.Fore.YELLOW}'
                         'No matching log directories found'
                         f'{colorama.Style.RESET_ALL}')
             return
 
-        job_ids = [log_dir[0] for log_dir in log_dirs]
+        job_ids = list(log_dirs.keys())
+        remote_log_dirs = list(log_dirs.values())
         local_log_dirs = [
```
[the remark this answers is not in the document]
Comment (romilbhardwaj) [reply sky/backends/cloud_vm_ray_backend.py:2080]: Ahh got it, thanks for clarifying!

**Review by romilbhardwaj — APPROVED** [review]
lgtm, thanks @Michaelvll! Should be good to merge after smoke tests. We should also add examples for the new APIs as a new PR

**Comment by Michaelvll** [comment]
Thank you for the review @romilbhardwaj! Just passed the smoke tests. Merging.

**Pull request merged**